panoply/datasource.py

```diff
@@ -20,6 +24,25 @@ def __init__(self, source, options={}):
         self.source = source
         self.options = options
 
+    @abstractmethod
+    def read(self, batch_size=None) -> List[PanoplyRecordGroup]:
+        """
+        Reads data from the sources and returns it as record group
+        """
+
+    @abstractmethod
+    def get_resource(self, source, options, resource_id: str) -> PanoplyResource:
+        """
+        Returns a PanoplyResource object with the list of fields
+        """
+
+    @abstractmethod
+    def list_resources(self, source, options) -> List[PanoplyResource]:
```

Comment:
if we need to validate the existence of these methods when we initialize the instance we should leave them inside the class (as in sdk we do not have access to the whole ve just to the Stream). But we still can use them as exports. The data sources will introduce such functionality.

Reply:
@yonivy I have a very strange approach but it helps to avoid initializing the class when the method is static.

1. __init__
```
from .ui_configuration import CONFIG
from .some_source import SomeSource as Stream
from .panoply_resources import list_resources, get_resource
```

2. panoply_resources.py
```
from .some_source import SomeSource


def init_instance(func_name, source, options, resource_name=None):
    try:
        function = getattr(SomeSource, func_name)
        return function(source, options, resource_name)
    except TypeError:
        print("Initializing instance of the class...")
        ds = SomeSource(source, options, validate=False)
        function = getattr(ds, func_name)
        return function(source, options, resource_name)


def list_resources(source, options):
    return init_instance("list_resources", source, options)


def get_resource(source, options, resource_id):
    return init_instance("get_resource", source, options, resource_id)
```

Reply:
How can this approach be used throughout the codebases? It seems like too many moving parts and therefore error prone. If you plan to have the functions exposed by the module then why bother adding the functions to the class?

Reply:
I can suggest two options:

1. Skip the validation in sdk and just use exports
**datasource**
**__init__**
```
from .ui_configuration import CONFIG
from .some_source import SomeSource as Stream
from .panoply_resources import list_resources, get_resource
```

**panoply_resources.py**
from panoply import Resource 
from .some_source import SomeSource


def list_resources(source, options):
    """static method"""
    return [Resource(resource) for resource in RESOURCES ]


def get_resource(source, options, resource_id):
     ds = SomeSource(source, options, resource_name)
    fields = ds.get_fields(resource_name)
    resource = Resource(...)
    return resource
```
**sources-runner**
```
def list_resources(source, options):
    return module.list_resources(source, options)


def get_resource(source, options, resource_id):
    return module(source, options, resource_id)
```

2. Leave methods inside class and add initialization to the sources-runner
**sources-runner**
```
def init_instance(stream, func_name, source, options, resource_name=None):
    try:
        function = getattr(stream, func_name)
        return function(source, options, resource_name)
    except TypeError:
        print("Initializing instance of the class...")
        ds = stream(source, options)
        function = getattr(ds, func_name)
        return function(source, options, resource_name)


def list_resources(source, options):
    return init_instance(module.Stream, "list_resources", source, options)


def get_resource(source, options, resource_id):
    return init_instance(module.Stream, "get_resource", source, options, resource_id)
```

@yonivy What do you think?

Reply:
btw, why can't we just do something like this in sources_runner?

```python
def list_resources(source, options):
    return module.Stream(source, options).list_resources()


def get_resource(source, options, resource_id):
    return module.Stream(source, options).get_resource(resource_id)
```

Reply:
oh, that's exactly what you wrote at the 2 option

Reply:
I can simply give a direction and move on, but I want the solution to be useful and friendly for the developers, so I'm relying greatly on your input here.

Here are my thoughts so far (in no particular order). Let's use them do find the best solution.

>PS if this is too much for a PR review lets zoom/slack about it

### As class methods
- Might be easier to get type hints - I need your knowledge on this (my python 3 is just ok)
- These methods can "bloat" the class with details that are needed only for configuration flows
- The caller must initialize the class

### As standalone functions
- More flexible as each source can do whatever they need. It is also possible to implement whatever's needed elsewhere and call it in the class, but it adds another level of indirection.
- How can we enforce an "export contract"? today both `CONFIG` and `Stream` are required by convention, but is there a python way to enforce it?
- Defining these functions in a dedicated submodule organizes things "nicely".

Reply:
@yonivy I think the best solution is to use class methods. 
Then we can do the following:

**datasource**

**__init__.py**
```
from .ui_configuration import CONFIG
from .some_source import SomeSource as Stream
```

**datasource.py**
```
from .constants import RESOURCES


class SomeSource:

    @classmethod
    def list_resources(cls, source, options):
        return [Resource(resource) for resource in RESOURCES ]
    
    @classmethod
    def get_resource(cls, source, options, resource_id):
        ds = SomeSource(source, options)
        fields = ds.get_fields(resource_id)
        resource = Resource(...)
        return resource
```

**sources-runner.py**
```
def list_resources(source, options):
    return module.Stream.list_resources(source, options)


def get_resource(source, options, resource_id):
    return module.Stream.get_resource(source, options, resource_id)
```

Reply:
Ok, lets go with class methods for now.